### PR TITLE
fix(api-docs): handle the default value of parameter correctly

### PR DESCRIPTION
**Pull request by iawia002**
**What this PR does / why we need it**:

For now, if we set a default value for a field of an auto parameter, the API docs generation will fail with the error below:

```
FATAL 0601-16:59:25.166+08 api.go:61 | json: error calling MarshalJSON for type spec.Swagger: json: error calling MarshalJSON for type *spec.Paths: json: error calling MarshalJSON for type spec.PathItem: json: error calling MarshalJSON for type *spec.Operation: json: error calling MarshalJSON for type spec.OperationProps: json: error calling MarshalJSON for type spec.Parameter: json: error calling MarshalJSON for type *swagger.rawJSON: invalid character 's' after top-level value
```

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #272 

**Special notes for your reviewer**:

/cc @hezhizhen @supereagle 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: utils/generators/swagger/generator.go [hunk 1]
```diff
@@ -495,14 +534,22 @@ func (g *Generator) enum(typ *api.Type) []spec.Parameter {
 		parameters := []spec.Parameter(nil)
 		if tag != "" {
 			source, name, apc, err := service.ParseAutoParameterTag(tag)
-			defaultValue := apc.Get(service.AutoParameterConfigKeyDefaultValue)
+			rawDefaultValue := apc.Get(service.AutoParameterConfigKeyDefaultValue)
+			var defaultValue []byte
+			if c := converterFor(string(field.Type)); rawDefaultValue != "" && c != nil {
+				// we didn't find a good way to handle the default value of none-basic types,
```
Comment (hezhizhen) [inline utils/generators/swagger/generator.go:540]: didn't? don't?

File: utils/generators/swagger/generator.go [hunk 2]
```diff
@@ -485,7 +487,44 @@ func (g *Generator) generateAutoParameter(typ api.TypeName) []spec.Parameter {
 		return nil
 	}
 	return g.enum(structType)
+}
 
+func converterFor(typ string) service.Converter {
```
Comment (hezhizhen) [inline utils/generators/swagger/generator.go:492]: 这个直接是 map 变量和一个函数有啥区别吗？

File: utils/generators/swagger/generator.go [hunk 3]
```diff
@@ -495,14 +534,22 @@ func (g *Generator) enum(typ *api.Type) []spec.Parameter {
 		parameters := []spec.Parameter(nil)
 		if tag != "" {
 			source, name, apc, err := service.ParseAutoParameterTag(tag)
-			defaultValue := apc.Get(service.AutoParameterConfigKeyDefaultValue)
+			rawDefaultValue := apc.Get(service.AutoParameterConfigKeyDefaultValue)
+			var defaultValue []byte
+			if c := converterFor(string(field.Type)); rawDefaultValue != "" && c != nil {
```
Comment (hezhizhen) [inline utils/generators/swagger/generator.go:539]: raw != nil && c == nil, default 就是空的吗？

File: utils/generators/swagger/generator.go (hunk 2 above)
Comment (iawia002) [reply utils/generators/swagger/generator.go:492]: 我再改一下，刚开始这个函数还有一些其它的东西

File: utils/generators/swagger/generator.go (hunk 3 above)
Comment (iawia002) [reply utils/generators/swagger/generator.go:539]: 对，空默认值，不直接报错

File: utils/generators/swagger/generator.go (hunk 1 above)
Comment (iawia002) [reply utils/generators/swagger/generator.go]: Updated

**Comment by iawia002** [comment]
All comments addressed, PTAL @hezhizhen

**Comment by hezhizhen** [comment]
/lgtm
but I'm not sure if it's expected...

**Comment by iawia002** [comment]
> but I'm not sure if it's expected...

What? Setting an empty default value for those non-basic types?

File: utils/generators/swagger/generator.go [hunk 4]
```diff
@@ -495,14 +531,22 @@ func (g *Generator) enum(typ *api.Type) []spec.Parameter {
 		parameters := []spec.Parameter(nil)
```
Comment (hezhizhen) [inline utils/generators/swagger/generator.go:531]: 这样子写的话，一开始长度就是 1？

Comment (iawia002) [reply utils/generators/swagger/generator.go:531]: 0

Comment (hezhizhen) [reply utils/generators/swagger/generator.go:531]: 噢，是小括号，不是大括号。

**Comment by iawia002** [comment]
If default values of non-basic types are not handled, the effect is as follows, the Swagger UI will not display the default values, that's all, I think it's better than returning errors directly during document generation.

Screen Shot 2020-06-02 at 12 06 44

So any more comments?

**Comment by hezhizhen** [comment]
> If default values of non-basic types are not handled, the effect is as follows, the Swagger UI will not display the default values, that's all, I think it's better than returning errors directly during document generation.
> 
> Screen Shot 2020-06-02 at 12 06 44
> 
> So any more comments?

works for me.

**Pull request merged**